{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Systems Immunogenetics Project\n",
    "\n",
    "## West Nile Virus QTL Random Walk Gene Prioritization\n",
    "\n",
    "### McWeeney Lab, Oregon Health & Science University\n",
    "\n",
    "#### Author: Steve Chamberlin (chambest@ohsu.edu)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "This document will run through steps that will use different sets of seed genes to perform a random walk with restarts analysis for a list of gene candidates identified from upstream QTL analyses done for this study. Gene candidates will be annotated if they are in the highest decile of the RWR scores, using a mouse protein-protein interaction network, indicating potential functional similarity to the seed gene sets. These seed gene sets were selected as described in the associated publication.\n",
    "\n",
    "Required Files:\n",
    "\n",
    "- This notebook** (`WNV_QTL_RandomWalk_Prioritization_Publication.ipynb`): [[Download here]](https://github.com/chambest/WNV_QTL_Resource/tree/main/Code)\n",
    "- The R script (contains required R packages and custom functions) (`WNV_rix_qtl_mapping_functions_publication.r`): [[Download here]](https://github.com/chambest/WNV_QTL_Resource/tree/main/Code)\n",
    "- Annotated variant dataframe necessary for this process (`allsnps_genesonly_final_infected.rda`): [[Download here]](https://figshare.com/articles/dataset/Processed_and_annotated_QTL_and_SNP_files/28147970)\n",
    "- STRING database to build the PPI network (`10090.protein.links.v11.0.txt`): [[Download here]](https://figshare.com/articles/dataset/String_files_used_to_created_the_PPI_used_in_the_RWR_analysis_in_the_paper/28147943)\n",
    "- JAX file to get mouse gene symbols (`PPI_MGIBatchReport_20210624_181229.xlsx`): [[Download here]](https://figshare.com/articles/dataset/String_files_used_to_created_the_PPI_used_in_the_RWR_analysis_in_the_paper/28147943)\n",
    "- Seed genes from iRNA study (File 1) (`NIHMS154918-supplement-Table_1_genes.csv`): [[Download here]](https://figshare.com/articles/dataset/Random_Walk_with_Restarts_Seed_Gene_Files/28147904)\n",
    "- Seed genes from iRNA study (File 2) (`NIHMS154918-supplement-Table_2_genes.csv`): [[Download here]](https://figshare.com/articles/dataset/Random_Walk_with_Restarts_Seed_Gene_Files/28147904)\n",
    "- T-cell receptor signalling pathway seed genes (Immport) (`TCR_Signaling_Pathway.txt`): [[Download here]](https://figshare.com/articles/dataset/Random_Walk_with_Restarts_Seed_Gene_Files/28147904)\n",
    "- Interferon seed genes (Immport) (`Interferons.txt`): [[Download here]](https://figshare.com/articles/dataset/Random_Walk_with_Restarts_Seed_Gene_Files/28147904)\n",
    "- Tumor necrosis factor family seed genes (Immport) (`TNF_Family_Members.txt`): [[Download here]](https://figshare.com/articles/dataset/Random_Walk_with_Restarts_Seed_Gene_Files/28147904)\n",
    "\n",
    "\n",
    "** Note: this notebook can also be downloaded as an R script (only the code blocks seen below will be included): [[Download R script here]](https://github.com/chambest/WNV_QTL_Resource/tree/main/Code)\n",
    "\n",
    "**All code is available on GitHub: (https://github.com/chambest/WNV_QTL_Resource/tree/main/Code) ** \n",
    "\n",
    "If you are not familiar with Jupyter Notebooks, here is the documentation online:\n",
    "\n",
    "1. [Jupyter for R Tutorial](http://nbviewer.jupyter.org/github/mooneymi/jupyter_notebooks/blob/master/r/Getting_Started_R.ipynb)\n",
    "2. [Jupyter Documentation](http://jupyter.org/)\n",
    "3. [Conda and R](https://www.continuum.io/conda-for-r)\n",
    "\n",
    "#### Output\n",
    "\n",
    "At the end, this program will create a dataframe containing the gene candidates annotated with a RWR score, as well as a score percentile within the entire STRING PPI, for each of the five separate gene sets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1. Load Necessary R Functions and Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load R functions and libraries and download required files \n",
    "source('WNV_rix_qtl_mapping_functions_publication.r')\n",
    "\n",
    "# Download allsnps_genesonly_final_infected.rda\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509954\", destfile = \"../DATA/allsnps_genesonly_final_infected.rda\")\n",
    "\n",
    "# Download 10090.protein.links.v11.0.txt\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509903\", destfile = \"../DATA/10090.protein.links.v11.0.txt\")\n",
    "\n",
    "# Download PPI_MGIBatchReport_20210624_181229.xlsx\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509900\", destfile = \"../DATA/PPI_MGIBatchReport_20210624_181229.xlsx\")\n",
    "\n",
    "# Download NIHMS154918-supplement-Table_1_genes.csv\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509840\", destfile = \"../DATA/NIHMS154918-supplement-Table_1_genes.csv\")\n",
    "\n",
    "# Download NIHMS154918-supplement-Table_2_genes.csv\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509825\", destfile = \"../DATA/NIHMS154918-supplement-Table_2_genes.csv\")\n",
    "\n",
    "# Download TCR_Signaling_Pathway.txt\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509822\", destfile = \"../DATA/TCR_Signaling_Pathway.txt\")\n",
    "\n",
    "# Download Interferons.txt\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509828\", destfile = \"../DATA/Interferons.txt\")\n",
    "\n",
    "# Download TNF_Family_Members.txt\n",
    "download.file(url = \"https://figshare.com/ndownloader/files/51509831\", destfile = \"../DATA/TNF_Family_Members.txt\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2. Load the annotated variant file and derive the list of candidate genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the master annotated variant file\n",
    "load(\"../DATA/allsnps_genesonly_final_infected.rda\")\n",
    "\n",
    "#### Select genes with high, moderate or deleterious variants matching the founder effects from the variant file, as defined in the associated publication:\n",
    "\n",
    "Mb15<-dplyr::filter(allsnps_genesonly_final_infected,(end.y-start.y>0 & end.y-start.y<=15 & finalcat != \"Cluster\"))\n",
    "\n",
    "highimpact<-filter(Mb15, (csq1==\"High Impact\" | csq1=='Moderate Impact' |\n",
    "                          SIFT_PREDICTION==\"DELETERIOUS\" | SIFT_PREDICTION==\"DELETERIOUS (*WARNING! Low confidence)\") &\n",
    "                         (finalcat==\"5\" | finalcat==\"10\")  )\n",
    "\n",
    "highimpactgenes<-distinct(highimpact,Marker.Symbol, genome.coordinate.start,genome.coordinate.end,chr)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3. Create the STRING PPI in iGRAPH and get human homologue genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read the base STRING interaction file and filter based on a physical interaction score>=700\n",
    "string<-read.delim('../DATA/10090.protein.links.v11.0.txt',stringsAsFactors=F,header=T, sep=\" \")\n",
    "string_real<-string[string$combined_score>700,]\n",
    "\n",
    "string_real$newprot1<-substr(string_real$protein1,str_locate(string_real$protein1,\"10090\")[,2]+2,length(string_real$protein1))\n",
    "string_real$newprot2<-substr(string_real$protein2,str_locate(string_real$protein2,\"10090\")[,2]+2,length(string_real$protein2))\n",
    "\n",
    "#Get the MGI symbol from the JAX file\n",
    "jax<-read.xls(\"../DATA/PPI_MGIBatchReport_20210624_181229.xlsx\")\n",
    "jax<-filter(jax,MGI.Gene.Marker.ID != 'No associated gene')\n",
    "\n",
    "string_real<-inner_join(string_real,dplyr::select(jax,Input,Symbol),by=c(\"newprot1\"=\"Input\"))\n",
    "string_real<-inner_join(string_real,dplyr::select(jax,Input,Symbol),by=c(\"newprot2\"=\"Input\"))\n",
    "\n",
    "string_real$Symbol.x<-toupper(string_real$Symbol.x)\n",
    "string_real$Symbol.y<-toupper(string_real$Symbol.y)\n",
    "\n",
    "string_uniq1<-distinct(string_real,Symbol.x) %>%\n",
    "              mutate(.,gene=Symbol.x) %>%\n",
    "              dplyr::select(.,-Symbol.x)\n",
    "string_uniq2<-distinct(string_real,Symbol.y) %>%\n",
    "              mutate(.,gene=Symbol.y) %>%\n",
    "              dplyr::select(.,-Symbol.y)\n",
    "\n",
    "uniq<-distinct(rbind(string_uniq1,string_uniq2),gene)\n",
    "\n",
    "#Create string PPI and perform various calculations\n",
    "library(igraph)\n",
    "string_ppi<-graph_from_data_frame(string_real[,c(6,7)],directed=FALSE)\n",
    "clusters<-components(string_ppi)\n",
    "clusters$csize\n",
    "clusters$no\n",
    "\n",
    "### Get human genes orthologues for the JAX mouse genes symbols\n",
    "library('biomaRt')\n",
    "ensembl <- useMart('ensembl', dataset=\"mmusculus_gene_ensembl\")\n",
    " \n",
    "human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\", host = \"https://dec2021.archive.ensembl.org/\")\n",
    "mouse = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\", host = \"https://dec2021.archive.ensembl.org/\")\n",
    "\n",
    "genesppi = getLDS(attributes = c(\"mgi_symbol\",\"ensembl_peptide_id\"), filters = \"mgi_symbol\", values = jax$Symbol , mart = mouse, attributesL = c(\"hgnc_symbol\"), martL = human, uniqueRows=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4. Annotate candidate gene file with RWR scores for the 305 iRNA seed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(RandomWalkRestartMH)\n",
    "\n",
    "# 305 iRNA seed genes for WNV infection involvement\n",
    "# Annotates the high impact genes with the percentile for the RWR score for this seed set\n",
    "\n",
    "seeds1<-read.csv(\"../DATA/NIHMS154918-supplement-Table_1_genes.csv\",header=T)\n",
    "seeds2<-read.csv(\"../DATA/NIHMS154918-supplement-Table_2_genes.csv\",header=T)\n",
    "seeds1$Symbol <- toupper(seeds1$Symbol)\n",
    "seeds2$Symbol <- toupper(seeds2$Symbol)\n",
    "\n",
    "#Get mouse gene homologues\n",
    "seedstemp1<-inner_join(seeds1,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"Symbol\"=\"HGNC.symbol\"))  \n",
    "seedstemp2<-inner_join(seeds2,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"Symbol\"=\"HGNC.symbol\"))  \n",
    "\n",
    "seedstemp1$MGI.symbol<-toupper(seedstemp1$MGI.symbol)\n",
    "seedstemp2$MGI.symbol<-toupper(seedstemp2$MGI.symbol)\n",
    "\n",
    "SeedGene1<-toupper(inner_join(seedstemp1,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol) \n",
    "SeedGene2<-toupper(inner_join(seedstemp2,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol) \n",
    "\n",
    "SeedGene<-c(SeedGene1,SeedGene2)\n",
    "SeedGene_305<-SeedGene\n",
    "\n",
    "PPI_Monoplex_Obj <- create.multiplex(list(string_ppi),Layers_Name=c(\"PPI\"))\n",
    "PPI_Monoplex_Obj\n",
    "\n",
    "AdjMatrix_PPI <- compute.adjacency.matrix(PPI_Monoplex_Obj)\n",
    "AdjMatrixNorm_PPI <- normalize.multiplex.adjacency(AdjMatrix_PPI)\n",
    "\n",
    "RWR_PPI_Results_305 <- Random.Walk.Restart.Multiplex(AdjMatrixNorm_PPI, PPI_Monoplex_Obj,SeedGene)\n",
    "\n",
    "seeddf<-data.frame(NodeNames=SeedGene,Score=rep(1,times=length(SeedGene)))\n",
    "df305<-rbind(RWR_PPI_Results_305$RWRM_Results,seeddf)\n",
    "\n",
    "highimpactgenes$Marker.Symbol<-toupper(highimpactgenes$Marker.Symbol)\n",
    "highimpactgenes<-left_join(highimpactgenes,dplyr::select(df305,NodeNames,Score),by=c(\"Marker.Symbol\"=\"NodeNames\"))\n",
    " \n",
    "highimpactgenes$pct305<-ecdf(df305$Score)(highimpactgenes$Score)  \n",
    "df305$pct305<-ecdf(df305$Score)(df305$Score) \n",
    "\n",
    "highimpactgenes$Score305<-highimpactgenes$Score\n",
    "highimpactgenes<-dplyr::select(highimpactgenes,-Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 5. Annotate candidate gene file with RWR scores for GeneWeaver West Nile virus seed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the GeneWeaver geneset for WNV\n",
    "# Annotates the high impact genes with the percentile for the RWR score for this seed set\n",
    "\n",
    "#WNV seed genes downloaded from the GeneWeaver website\n",
    "SeedGenedf <- data.frame(gene=c(\"BCL2L12\",\n",
    "              \"CCR5\",\n",
    "              \"CXCL10\",\n",
    "              \"CXCR3\",\n",
    "              \"DDX58\",\n",
    "              \"EIF2AK2\",\n",
    "              \"IFIT1\",\n",
    "              \"IFIT2\",\n",
    "              \"IFNAR1\",\n",
    "              \"IFNB1\",\n",
    "              \"IFNG\",\n",
    "              \"IRF3\",\n",
    "              \"IRF7\",\n",
    "              \"MAVS\",\n",
    "              \"MMP9\",\n",
    "              \"MYD88\",\n",
    "              \"OAS1\",\n",
    "              \"PRF1\",\n",
    "              \"RAG1\",\n",
    "              \"TLR7\",\n",
    "              \"TNF\"))\n",
    "\n",
    "#Get mouse gene homologues\n",
    "seedstemp<-inner_join(SeedGenedf,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"gene\"=\"HGNC.symbol\"))\n",
    "seedstemp$MGI.symbol<-toupper(seedstemp$MGI.symbol)\n",
    "SeedGene<-toupper(inner_join(seedstemp,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol)\n",
    "\n",
    "SeedGene_GW<-SeedGene\n",
    "\n",
    "RWR_PPI_Results_DW <- Random.Walk.Restart.Multiplex(AdjMatrixNorm_PPI, PPI_Monoplex_Obj,SeedGene)\n",
    " \n",
    "seeddf<-data.frame(NodeNames=SeedGene,Score=rep(1,times=length(SeedGene)))\n",
    "dfDW<-rbind(RWR_PPI_Results_DW$RWRM_Results,seeddf)\n",
    "\n",
    "highimpactgenes<-left_join(highimpactgenes,dplyr::select(dfDW,NodeNames,Score),by=c(\"Marker.Symbol\"=\"NodeNames\"))\n",
    "\n",
    "highimpactgenes$pctDW<-ecdf(dfDW$Score)(highimpactgenes$Score)  \n",
    "dfDW$pctDW<-ecdf(dfDW$Score)(dfDW$Score)  \n",
    "\n",
    "highimpactgenes$ScoreGW<-highimpactgenes$Score\n",
    "highimpactgenes<-dplyr::select(highimpactgenes,-Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6. Annotate candidate gene file with RWR scores for the t-cell receptor signaling pathway seed genes from Immport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Immport random walk for t-cell receptor signaling pathway genes\n",
    "# Annotates the high impact genes with the percentile for the RWR score for this seed set\n",
    "\n",
    "tcr<-read.table('../DATA/TCR_Signaling_Pathway.txt',header=T,sep=\"\\t\") \n",
    "tcrgenes<-dplyr::distinct(tcr,Symbol)\n",
    "\n",
    "#Get mouse gene homologues\n",
    "seedstemp<-inner_join(tcrgenes,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"Symbol\"=\"HGNC.symbol\"))\n",
    "seedstemp$MGI.symbol<-toupper(seedstemp$MGI.symbol)\n",
    "SeedGene<-toupper(inner_join(seedstemp,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol)\n",
    "\n",
    "SeedGene_tcr<-SeedGene\n",
    "\n",
    "RWR_PPI_Results_tcr <- Random.Walk.Restart.Multiplex(AdjMatrixNorm_PPI, PPI_Monoplex_Obj,SeedGene)\n",
    "\n",
    "seeddf<-data.frame(NodeNames=SeedGene,Score=rep(1,times=length(SeedGene)))\n",
    "dftcr<-rbind(RWR_PPI_Results_tcr$RWRM_Results,seeddf)\n",
    "\n",
    "highimpactgenes<-left_join(highimpactgenes,dplyr::select(dftcr,NodeNames,Score),by=c(\"Marker.Symbol\"=\"NodeNames\"))\n",
    "\n",
    "highimpactgenes$pcttcr<-ecdf(dftcr$Score)(highimpactgenes$Score)  \n",
    "dftcr$pcttcr<-ecdf(dftcr$Score)(dftcr$Score) \n",
    "\n",
    "highimpactgenes$Scoretcr<-highimpactgenes$Score\n",
    "highimpactgenes<-dplyr::select(highimpactgenes,-Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 7. Annotate candidate gene file with RWR scores for the interferon seed genes from Immport."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Immport random walk for interferon genes\n",
    "# Annotates the high impact genes with the percentile for the RWR score for this seed set\n",
    "\n",
    "ifn<-read.table('../DATA/RWR/Interferons.txt',header=T,sep=\"\\t\") \n",
    "ifngenes<-dplyr::distinct(ifn,Symbol)\n",
    "\n",
    "#Get mouse gene homologues\n",
    "seedstemp<-inner_join(ifngenes,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"Symbol\"=\"HGNC.symbol\"))\n",
    "seedstemp$MGI.symbol<-toupper(seedstemp$MGI.symbol)\n",
    "SeedGene<-toupper(inner_join(seedstemp,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol)\n",
    "\n",
    "SeedGene_ifn<-SeedGene\n",
    "\n",
    "RWR_PPI_Results_ifn <- Random.Walk.Restart.Multiplex(AdjMatrixNorm_PPI, PPI_Monoplex_Obj,SeedGene)\n",
    "\n",
    "seeddf<-data.frame(NodeNames=SeedGene,Score=rep(1,times=length(SeedGene)))\n",
    "dfifn<-rbind(RWR_PPI_Results_ifn$RWRM_Results,seeddf)\n",
    "\n",
    "highimpactgenes<-left_join(highimpactgenes,dplyr::select(dfifn,NodeNames,Score),by=c(\"Marker.Symbol\"=\"NodeNames\"))\n",
    "\n",
    "highimpactgenes$pctifn<-ecdf(dfifn$Score)(highimpactgenes$Score)  \n",
    "dfifn$pctifn<-ecdf(dfifn$Score)(dfifn$Score)  \n",
    " \n",
    "highimpactgenes$Scoreifn<-highimpactgenes$Score\n",
    "highimpactgenes<-dplyr::select(highimpactgenes,-Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 8. Annotate candidate gene file with RWR scores for the tumor necrosis factor seed genes from Immport."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Immport random walk for tfn genes\n",
    "# Annotates the high impact genes with the percentile for the RWR score for this seed set\n",
    "\n",
    "tnf<-read.table('../DATA/TNF_Family_Members.txt',header=T,sep=\"\\t\") \n",
    "tnfgenes<-dplyr::distinct(tnf,Symbol)\n",
    "\n",
    "#Get mouse gene homologues\n",
    "seedstemp<-inner_join(tnfgenes,distinct(genesppi,HGNC.symbol,MGI.symbol),by=c(\"Symbol\"=\"HGNC.symbol\"))\n",
    "seedstemp$MGI.symbol<-toupper(seedstemp$MGI.symbol)\n",
    "SeedGene<-toupper(inner_join(seedstemp,uniq,by=c(\"MGI.symbol\"=\"gene\"))$MGI.symbol)\n",
    "\n",
    "SeedGene_tnf<-SeedGene\n",
    "\n",
    "RWR_PPI_Results_tnf <- Random.Walk.Restart.Multiplex(AdjMatrixNorm_PPI, PPI_Monoplex_Obj,SeedGene)\n",
    "\n",
    "seeddf<-data.frame(NodeNames=SeedGene,Score=rep(1,times=length(SeedGene)))\n",
    "dftnf<-rbind(RWR_PPI_Results_tnf$RWRM_Results,seeddf)\n",
    "\n",
    "highimpactgenes<-left_join(highimpactgenes,dplyr::select(dftnf,NodeNames,Score),by=c(\"Marker.Symbol\"=\"NodeNames\"))\n",
    "\n",
    "highimpactgenes$pcttnf<-ecdf(dftnf$Score)(highimpactgenes$Score)  \n",
    "dftnf$pcttnf<-ecdf(dftnf$Score)(dftnf$Score)  \n",
    "\n",
    "highimpactgenes$Scoretnf<-highimpactgenes$Score\n",
    "highimpactgenes<-dplyr::select(highimpactgenes,-Score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(highimpactgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
